### PR TITLE
[TwigBundle] Refresh twig paths when resources change.

### PR DESCRIPTION
**Pull request by aitboudad**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Fixed tickets | ~ |
| Tests pass? | yes |
| License | MIT |

### CODE REVIEW

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php [hunk 1]
```diff
@@ -77,19 +78,27 @@ public function load(array $configs, ContainerBuilder $container)
         }
 
         // register bundles as Twig namespaces
+        $dirs = array();
         foreach ($container->getParameter('kernel.bundles') as $bundle => $class) {
             if (is_dir($dir = $container->getParameter('kernel.root_dir').'/Resources/'.$bundle.'/views')) {
                 $this->addTwigPath($twigFilesystemLoaderDefinition, $dir, $bundle);
+                $dirs[] = $dir;
```
Comment (xabbuh) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: Why don't we call `addResource()` here directly?

Comment (aitboudad) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: just to  avoid call ```$container->addResource(new DirectoryResource($dir));``` in several places

Comment (stof) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: why avoiding it ?

Comment (aitboudad) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: fixed 2 vs 1 :), seem there is not big difference but I still like the old way.

**Comment by xabbuh** [comment]
:+1:

**Comment by stloyd** [comment]
:+1:

**Comment by aitboudad** [comment]
ping @fabpot

**Comment by fabpot** [comment]
Why do you consider this as a new feature? Looks like a bug to me, no?

**Comment by aitboudad** [comment]
@fabpot I think so as it throw an exception when we remove an already registed path, you can merge it into `2.3` without problems

**Comment by fabpot** [comment]
Thank you @aitboudad.

**Comment by chbruyand** [comment]
Hello,

This commit has a *huge* (+900%) impact on performances when the resources are being tracked by the container.
Is this really worth the benefits it gives ?

In fact, I don't understand why the Resources/view directories are added to the tracked resources ?

Regards.

**Comment by stof** [comment]
@chbruyand because we need to reconfigure twig when the directory does not exist anymore.

@aitboudad @fabpot I think we actually need a ``DirectoryExistenceResource`` here, which would check for freshness in a different way. We don't care about the modification date of the directory for these cases. And we care about a case not covered by the current implementation: adding new directories in a place not detected previously (which still requires to clear the cache manually in the current setup). And actually, many other places using DirectoryResource currently should actually use such DirectoryExistenceResource (translator configuration paths or DoctrineBundle mapping paths for instance)

@chbruyand does this have an impact on all page load or only the first time after the symfony update ? If it is only the first time, it might be becaus eyou compare a request rebuilding the cache with a request not rebuilding it.

**Comment by chbruyand** [comment]
@stof this is on all requests being made, not only when building the container for the first time.

**Comment by mpdude** [comment]
@stof Do I get you right that in this case the contents of the directory do not matter, just the fact that a particular directory *does* or *does not* exist with the resource being stale if that changes?

**Comment by reisba** [comment]
I updated my dev-environment to symfony 2.3.29 which seems to contain this patch and it really seems to have a performance penalty in the figures that @chbruyand pointed out. What makes it worse is that the site I'm writing makes additional ajax requests to the symfony-backend after the initial page-load and each one of the seems to be impacted in the same way.

I don't know if there's issues with my environment configuration, but going from .28 to .29 really broke the performance for me in dev-environment and these changes seem to be at the bottom of it for me.

**Comment by chbruyand** [comment]
@stof I was starting to work on a patch based on your thoughts. Using a FileResource instead of a DirectoryResource would ensure twig is reconfigured when directory is deleted (I think there is no need for a DirectoryExistenceResource there). If we really want to skip the modification time check in that case, I would argue in favour of a FileExistenceResource or an extra argument to the existing FileResource.

**Comment by stof** [comment]
@chbruyand we would also need to have it reconfigured when the non-existent directory is added though

**Comment by stof** [comment]
And yes, it can be named ``FileExistenceResource``, working with both files and directories.

**Comment by chbruyand** [comment]
@stof I made a pull request (see #14773), comments are welcome

**Pull request closed without merging**